src/translation.cpp

```diff
@@ -145,11 +211,337 @@ void Translations::loadTranslation(const std::string &data)
+						c = c * 16 + v;
+					} else {
+						errorstream << "Invalid unicode escape sequence \"\\u" << wide_to_utf8(str.substr(i, 4)) << "\", ignoring" << std::endl;
+						ok = false;
+						break;
+					}
+				}
+				if (ok) {
+					wide_add_codepoint(result, c);
+				}
+				i += 4;
+				continue;
+			}
+			case L'U': {
+				i++;
+				if (i + 8 > str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char32_t c = 0;
+				bool ok = true;
+				for (size_t j = 0; j < 8; j++) {
+					unsigned char v;
+					if (str[i+j] <= 0xff && hex_digit_decode((char)str[i+j], v)) {
+						c = c * 16 + v;
+					} else {
+						errorstream << "Invalid unicode escape sequence \"\\U" << wide_to_utf8(str.substr(i, 8)) << "\", ignoring" << std::endl;
+						ok = false;
+						break;
+					}
+				}
+				if (ok) {
+					wide_add_codepoint(result, c);
+				}
+				i += 8;
+				continue;
+			}
+			default: {
+				errorstream << "Unknown escape sequence \"\\" << str[i] << "\", ignoring" << std::endl;
+				break;
+			}
+		}
+		i++;
+	}
+	return result;
+}
+
+void Translations::loadPoEntry(const std::map<std::wstring, std::wstring> &entry) {
+	// Process an entry from a PO file and add it to the translation table
+	// Assumes that entry[L"msgid"] is always defined
+	std::wstring textdomain;
+	auto ctx = entry.find(L"msgctxt");
+	if (ctx != entry.end()) {
+		textdomain = ctx->second;
+	} else {
+		textdomain = L"";
+	}
+	std::wstring original = entry.at(L"msgid");
+
+	auto plural = entry.find(L"msgid_plural");
+	if (plural == entry.end()) {
+		auto translated = entry.find(L"msgstr");
+		if (translated == entry.end()) {
+			errorstream << "Could not load translation: entry for msgid \"" << wide_to_utf8(original) << "\" does not contain a msgstr field" << std::endl;
+			return;
+		}
+		addTranslation(textdomain, original, translated->second);
+	} else {
+		std::vector<std::wstring> translations;
+		for (int i = 0; ; i++) {
+			auto translated = entry.find(L"msgstr[" + std::to_wstring(i) + L"]");
+			if (translated == entry.end()) break;
+			translations.push_back(translated->second);
+		}
+		if (translations.size() == 0) {
+			errorstream << "Could not load translation: entry for msgid\"" << wide_to_utf8(original) << "\" does not contain a msgstr[0] field" << std::endl;
+			return;
+		}
+		addPluralTranslation(textdomain, original, translations);
+		addPluralTranslation(textdomain, plural->second, translations);
+	}
+}
+
+void Translations::loadPoTranslation(const std::string &data) {
+	std::istringstream is(data);
+	std::string line;
+	std::map<std::wstring, std::wstring> last_entry;
+	std::wstring last_key;
+
+	while (is.good()) {
+		std::getline(is, line);
+		// Trim last character if file was using a \r\n line ending
+		if (line.length () > 0 && line[line.length() - 1] == '\r')
+			line.resize(line.length() - 1);
+
+		if (line.empty() || line[0] == '#')
+			continue;
```

Comment:
Since `xgettext` can generate [fuzzy entries](https://www.gnu.org/software/gettext/manual/html_node/Fuzzy-Entries.html) it would make sense to check for (and skip) those.

Reply:
From the documentation of gettext, it seems to say that fuzzy entries are considered as normal entries for most purposes, so I wouldn't drop them. In particular, the `.po` to `.mo` compilation would use them, so we should use them as well if we don't want to introduce a discrepancy.

Reply:
> From the documentation of gettext, it seems to say that fuzzy entries are considered as normal entries for most purposes, so I wouldn't drop them.

While that is the case, the problem with fuzzy matching is that it is enabled by default in `msgmerge` (unless the `-N` flag is passed) and can often match entries with different (sometimes opposite, e.g. "enabled" and "disabled") meanings while assigning the same translation if one is missing, and there are also reasons to generate fuzzy entries, e.g. if the source string does not change much (adding a period to the end of the source string) or to give translators a base string to modify/work with.

> In particular, the `.po` to `.mo` compilation would use them, so we should use them as well if we don't want to introduce a discrepancy.

`msgfmt` does not seem to include fuzzy entries by default (there is the [`-f` flag](https://www.gnu.org/software/gettext/manual/html_node/msgfmt-Invocation.html#Input-file-interpretation-1) option for including them which GNU does not recommend using).

Reply:
Thanks for the documentation! I'll implement matching the fuzzy entries in that case.